serde-generate/src/rust.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) Facebook, Inc. and its affiliates
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use crate::analyzer;
+use serde_reflection::{ContainerFormat, Format, Named, Registry, VariantFormat};
+use std::collections::{BTreeMap, HashSet};
+use std::io::{Result, Write};
+
+pub fn output(
+    out: &mut dyn Write,
+    with_derive_macros: bool,
+    registry: &Registry,
+) -> std::result::Result<(), Box<dyn std::error::Error>> {
+    let dependencies = analyzer::get_dependency_map(registry)?;
+    let entries = analyzer::best_effort_topological_sort(&dependencies)?;
+
+    output_preambule(out, with_derive_macros)?;
+    let mut known_sizes = HashSet::new();
+    for name in entries {
+        let format = &registry[name];
+        output_container(out, with_derive_macros, name, format, &known_sizes)?;
+        known_sizes.insert(name);
+    }
+    Ok(())
+}
+
+fn output_preambule(out: &mut dyn Write, with_derive_macros: bool) -> Result<()> {
+    let serde_imports = if with_derive_macros {
+        r#"
+use serde::{Serialize, Deserialize};
+use serde_bytes::ByteBuf;"#
+    } else {
+        // for testing
+        "type ByteBuf = Vec<u8>;"
+    };
+    writeln!(
+        out,
+        r#"
+#![allow(unused_imports)]{}
+use std::collections::BTreeMap;
+"#,
+        serde_imports
+    )
+}
+
+fn quote_type(format: &Format, known_sizes: Option<&HashSet<&str>>) -> String {
+    use Format::*;
+    match format {
+        TypeName(x) => {
+            if let Some(set) = known_sizes {
+                if !set.contains(x.as_str()) {
+                    return format!("Box<{}>", x);
+                }
+            }
+            x.to_string()
+        }
+        Unit => "()".into(),
+        Bool => "bool".into(),
+        I8 => "i8".into(),
+        I16 => "i16".into(),
+        I32 => "i32".into(),
+        I64 => "i64".into(),
+        I128 => "i128".into(),
+        U8 => "u8".into(),
+        U16 => "u16".into(),
+        U32 => "u32".into(),
+        U64 => "u64".into(),
+        U128 => "u128".into(),
+        F32 => "f32".into(),
+        F64 => "f64".into(),
+        Char => "char".into(),
+        Str => "String".into(),
+        Bytes => "ByteBuf".into(),
+
+        Option(format) => format!("Option<{}>", quote_type(format, known_sizes)),
+        Seq(format) => format!("Vec<{}>", quote_type(format, None)),
+        Map { key, value } => format!(
+            "BTreeMap<{}, {}>",
```

Comment:
Does this have to be  BTreeMap?

Reply:
Well, having a stable ordering is definitely more convenient for testing with bincode. I suppose we can always add an option later if needed.